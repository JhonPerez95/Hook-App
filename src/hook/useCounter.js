import { useState } from 'react';

const useCounter = (initialState = 10) => {
  const [state, setState] = useState(initialState);

  const increment = (value = 1) => {
    setState(state + value);
  };
  const decrement = (value = 1) => {
    setState(state - value);
  };

  const resertState = () => {
    setState(initialState);
  };

  return {
    increment,
    decrement,
    resertState,
    state,
  };
};

export default useCounter;
