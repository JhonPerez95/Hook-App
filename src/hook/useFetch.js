import { useState, useEffect, useRef } from 'react';

export const useFetch = (url) => {
  const isMounted = useRef(true);

  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url)
      .then((res) => res.json())
      .then((data) => {
        if (isMounted.current) {
          setState({
            data,
            loading: false,
            error: null,
          });
        } else {
          console.log('setState no se llamo');
        }
      })
      .catch((err) => {
        setState({
          data: null,
          loading: true,
          error: 'No se pudo cargar la info',
        });
      });
  }, [url]);

  return state;
};
