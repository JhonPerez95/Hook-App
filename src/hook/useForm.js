import { useState } from 'react';

const useForm = (initialState = {}) => {
  const [value, setValue] = useState(initialState);

  const handleInputChange = ({ target }) => {
    setValue({ ...value, [target.name]: target.value });
  };

  return [value, handleInputChange];
};

export default useForm;
