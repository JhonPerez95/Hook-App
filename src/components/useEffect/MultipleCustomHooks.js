import React from 'react';
import { useFetch } from '../../hook/useFetch';
import useCounter from '../../hook/useCounter';

const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(2);

  const url = `https://www.breakingbadapi.com/api/quotes/${counter}`;
  const { loading, data } = useFetch(url);
  const { author, quote } = !!data && data[0];

  //
  return (
    <div>
      <h1> Custom Hooks </h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">loading ...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-0">{quote}</p>
          <footer className="blockquote-footer">{author}</footer>
        </blockquote>
      )}
      <button type="button" className="btn btn-success" onClick={increment}>
        Next Quote
      </button>
    </div>
  );
};

export default MultipleCustomHooks;
