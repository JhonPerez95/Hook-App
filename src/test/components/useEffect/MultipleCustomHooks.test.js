import React from 'react';
import { shallow } from 'enzyme';

jest.mock('../../../hook/useFetch');

import MultipleCustomHooks from '../../../components/useEffect/MultipleCustomHooks';
import { useFetch } from '../../../hook/useFetch';

describe('Pruebas al component <MultipleCustomHooks/>', () => {
  test('Debe mostrarse correctamente', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe mostrar la informacion', () => {
    useFetch.mockReturnValue({
      data: [{ author: 'jhon', quote: 'Hola Mundo' }],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    // console.log(wrapper.html());

    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
    expect(wrapper.find('footer').text().trim()).toBe('jhon');
  });
});
