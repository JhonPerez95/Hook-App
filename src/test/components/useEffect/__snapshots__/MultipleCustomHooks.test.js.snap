// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al component <MultipleCustomHooks/> Debe mostrarse correctamente 1`] = `
<div>
  <h1>
     Custom Hooks 
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading ...
  </div>
  <button
    className="btn btn-success"
    onClick={[Function]}
    type="button"
  >
    Next Quote
  </button>
</div>
`;
